fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,127 @@
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => {
+        newTodo = this.state.todos;
+        data = newTodo.concat(data);
+        this.setState({ todos: data });
+      });
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const { todos } = this.state;
+    const search = index => index.todoId === todoId;
+    const index = todos.findIndex(search);
+    const indexId = todos[index].todoId;
+    const toggleCompleted = {
+      isCompleted: !todos[index].isCompleted
+    };
+
+    if (todoId === indexId) {
+
+      todos[index].isCompleted
+        ? todos[index] = toggleCompleted
+        : todos[index] = toggleCompleted;
+
+      fetch(`/api/todos/${todoId}`, {
+        method: 'PATCH',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(todos[index])
+      })
+        .then(response => response.json())
+        .then(data => {
+          const newTodos = todos;
+          newTodos[index] = data;
```

Comment:
This will also mutate the state, we need to make a copy of the array then update it. There are several ways to make a copy of an array but here are 2 examples:

Using `slice`:
```JSX
const newTodos = todos.slice();
newTodos[index] = data;
```

OR Using es6 spread operator:
```JSX
const newTodos = [ ...todos ];
newTodos[index] = data;
```